// @notenoughupdates/discord-akairo/dist/src/struct/AkairoClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AkairoClient = void 0;
const discord_js_1 = require("discord.js");
const ClientUtil_js_1 = require("./ClientUtil.js");
/**
 * The Akairo framework client. Creates the handlers and sets them up.
 */
class AkairoClient extends discord_js_1.Client {
    constructor(options, clientOptions) {
        const combinedOptions = { ...options, ...clientOptions };
        super(combinedOptions);
        this.ownerID = combinedOptions.ownerID ?? [];
        this.superUserID = combinedOptions.superUserID ?? [];
        this.util = new ClientUtil_js_1.ClientUtil(this);
    }
    /**
     * Checks if a user is the owner of this bot.
     * @param user - User to check.
     */
    isOwner(user) {
        const id = this.users.resolveId(user);
        if (!id)
            return false;
        return Array.isArray(this.ownerID) ? this.ownerID.includes(id) : id === this.ownerID;
    }
    /**
     * Checks if a user is a super user of this bot.
     * @param user - User to check.
     */
    isSuperUser(user) {
        const id = this.users.resolveId(user);
        if (!id)
            return false;
        return Array.isArray(this.superUserID)
            ? this.superUserID.includes(id) || this.ownerID.includes(id)
            : id === this.superUserID || id === this.ownerID;
    }
}
exports.AkairoClient = AkairoClient;
